sympy/polys/series/tests/test_powerseriesring.py

```diff
@@ -0,0 +1,257 @@
+    assert R0.equal_repr(R0.zero, R0.one)
+    assert R0.pretty(gen0) == "O(x**0)"
+
+    assert same(R0, gen0, [], 0)
+    assert same(R0, R0.multiply(gen0, gen0), [], 0)
+    assert same(R, R.from_list([1, 2, 3, 4, 5, 6, 7]), [1, 2, 3, 4, 5, 6], 6)
+    assert same(R, R.add(R.from_list([2, 4, 5], 3), R.from_list([5], 2)), [7, 4], 2)
+
+
+@pytest.mark.parametrize("r", list(Ring_ZZ + Ring_QQ))
+def test_int(r):
+    R = r()
+    x = R.gen
+    one = R.one
+    zero = R.zero
+
+    R3 = r(3)
+    x3 = R3.gen
+    one3 = R3.one
+
+    R10 = r(10)
+    x10 = R10.gen
+    one10 = R10.one
+
+    assert same(R, zero, [], None)
+    assert same(R, one, [1], None)
+    assert same(R, x, [0, 1], None)
+
+    assert same(R, R.negative(R.add(x, one)), [-1, -1], None)
+    assert same(R3, R3.add(x3, one3), [1, 1], None)
+    assert same(R3, R3.add(one3, R3.pow_int(x3, 4)), [1], 3)
+
+    assert same(R, R.subtract(R.multiply(x, x), x), [0, -1, 1], None)
+    assert same(R3, R3.subtract(R3.pow_int(R3.add(one3, x3), 4), x3), [1, 3, 6], 3)
+
+    assert same(R, R.multiply(x, x), [0, 0, 1], None)
+    assert same(
+        R3, R3.multiply(R3.square(R3.add(x3, one3)), R3.add(x3, one3)), [1, 3, 3], 3
+    )
+    assert same(
+        R10, R10.multiply(R10.add(x10, one10), R10.add(x10, one10)), [1, 2, 1], None
+    )
+
+    assert same(R, R.multiply_ground(one, 1), [1], None)
+    assert same(R, R.multiply_ground(x, -1), [0, -1], None)
+    assert same(R, R.multiply_ground(x, 0), [], None)
+    assert same(R, R.multiply_ground(R.square(x), 1), [0, 0, 1], None)
+    assert same(R, R.multiply_ground(R.add(x, one), ZZ(3)), [3, 3], None)
+
+    assert not same(R, R.square(R.add(x, one)), [1, 1, 1], None)
+    assert same(R3, R3.square(R3.multiply(x3, R3.add(x3, one3))), [0, 0, 1], 3)
+
+    assert same(R, R.pow_int(x, 0), [1], None)
+    assert same(R, R.pow_int(R.add(x, one), 6), [1, 6, 15, 20, 15, 6], 6)
+    assert same(R, R.pow_int(x, 10), [], 6)
+    assert same(R3, R3.pow_int(R3.add(x3, one3), 5), [1, 5, 10], 3)
+    assert same(R10, R10.pow_int(x10, 7), [0, 0, 0, 0, 0, 0, 0, 1], None)
+    assert same(
+        R10,
+        R10.pow_int(R10.add(x10, one10), 12),
+        [1, 12, 66, 220, 495, 792, 924, 792, 495, 220],
+        10,
+    )
+
+    assert same(R, R.truncate(R.pow_int(x, 3), 4), [0, 0, 0, 1], None)
+    assert same(R, R.truncate(R.pow_int(R.add(x, one), 5), 3), [1, 5, 10], 3)
+
+    assert same(R, R.differentiate(R.pow_int(x, 3)), [0, 0, 3], None)
+    assert same(R, R.differentiate(R.add(R.multiply(x, x), x)), [1, 2], None)
+    assert same(R3, R3.differentiate(R3.multiply(x3, x3)), [0, 2], None)
+    assert same(R3, R3.differentiate(R3.add(x3, one3)), [1], None)
+    assert same(R10, R10.differentiate(R10.pow_int(x10, 4)), [0, 0, 0, 4], None)
+    assert same(
+        R10, R10.differentiate(R10.add(R10.multiply(x10, x10), x10)), [1, 2], None
+    )
+
+
+@pytest.mark.parametrize("r", Ring_QQ)
+def test_rational(r):
```

Comment:
This is a good way to parametrise tests. This and the other test are a little long though.  I would maybe split them into smaller tests.